src/lib/one-graph/cli-netlify-graph.js

```diff
@@ -545,7 +679,185 @@ const loadNetlifyGraphConfig = (siteRoot) => {
+
+    if (newModule) {
+      const hasGenerators = Array.isArray(newModule.generators)
+      let generatorsConform = true
+      if (hasGenerators) {
+        newModule.generators.forEach((generator) => {
+          const hasId = Boolean(generator.id)
+          const hasName = Boolean(generator.name)
+          const hasVersion = Boolean(generator.version)
+          const hasGenerator = Boolean(generator.generateHandler)
+
+          if (!hasId || !hasName || !hasVersion || !hasGenerator) {
+            generatorsConform = false
+          }
+        })
+      }
+
+      if (!generatorsConform) {
+        warn(`Generator ${importPath} does not conform to expected module declaration type, ignoring...`)
+        return
+      }
+    }
+
+    return newModule
+  } catch (error_) {
+    warn(`Failed to load Graph code generator, please make sure that ${importPath} is either exists as a local file or is listed in your package.json under devDependencies, and that it's a commonjs (not esm) dependency.
+
+${error_}`)
+  }
+}
+
+const getCodegenModule = ({ config }) => dynamicallyLoadCodegenModule({ config })
+
+const getCodegenFunctionById = async ({ config, id }) => {
+  const codegenModule = await getCodegenModule({ config })
+
+  return codegenModule && codegenModule.generators && codegenModule.generators.find((generator) => generator.id === id)
+}
+
+const autocompleteCodegenModules = async ({ config }) => {
+  const codegenModule = await getCodegenModule({ config })
+  if (!codegenModule || !codegenModule.generators) {
+    return null
+  }
+
+  log(`Using Graph Codegen module ${codegenModule.id} [${codegenModule.version}] from '${config.graph.codeGenerator}'`)
+
+  const allGeneratorChoices = codegenModule.generators
+    // eslint-disable-next-line id-length
+    .sort((a, b) => a.name.localeCompare(b.name))
+    .map((codeGen) => ({
+      name: `${codeGen.name} [${codeGen.id}]`,
+      value: codeGen.name,
+    }))
+
+  // eslint-disable-next-line unicorn/consistent-function-scoping
+  const filterModuleNames = (moduleChoices, input) =>
+    moduleChoices.filter((moduleChoice) => moduleChoice.name.toLowerCase().match(input.toLowerCase()))
+
+  // eslint-disable-next-line n/global-require
+  const inquirerAutocompletePrompt = require('inquirer-autocomplete-prompt')
+  /** multiple matching detectors, make the user choose */
+  // @ts-ignore
+  inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt)
+
+  const perPage = 50
+
+  // @ts-ignore
+  const { selectedCodeGen } = await inquirer.prompt({
+    name: 'selectedCodeGen',
+    message: `Which codegen would you like to use?`,
+    type: 'autocomplete',
+    pageSize: perPage,
+    source(_, input) {
+      if (!input || input === '') {
+        return allGeneratorChoices
+      }
+
+      const filteredChoices = filterModuleNames(allGeneratorChoices, input)
+      // only show filtered results
+      return filteredChoices
+    },
+  })
+
+  return codegenModule.generators.find(
```

Comment:
maybe do a `|| []` here?

Reply:
I just checked, and TypeScript complains appropriately if this check isn't here https://github.com/netlify/cli/pull/4896/files/131a5f032bcf9a6e77fee98e273541ec5cbdeccd#diff-b50e55f8472633b964d2a9ce70b737d017c52297dfcf2be9ea1a11e9f7c70ad2R810